### PR TITLE
Refactoring the embedding intent classifier

### LINKED ISSUE #1412: refactor tf nlu classifier according to embedding_policy from rasa_core
Rasa_core's embedding policy is basically build on top of tf nlu classifier, but its code was refactored and clean-up. So tf nlu classifier should be refactored in similar fashion, mainly its config parameters for `hidden_layers`

### PR DESCRIPTION
**Pull request by MetcalfeTom**
Fixes #1412

**Proposed changes**:
- Refashion config parameters
- Re-structure functions within the classifier class to mimic the embedding policy class from Rasa Core
- Add comments & types
- Apply blusher & contour to bring out those gorgeous cheekbones

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

File: docs/components.rst [hunk 1]
```diff
@@ -327,8 +327,8 @@ intent_classifier_tensorflow_embedding
 
     The algorithm also has hyperparameters to control:
         - neural network's architecture:
-            - ``num_hidden_layers_a`` and ``hidden_layer_size_a`` set the number of hidden layers and their sizes before embedding layer for user inputs;
-            - ``num_hidden_layers_b`` and ``hidden_layer_size_b`` set the number of hidden layers and their sizes before embedding layer for intent labels;
+            - ``hidden_layers_sizes_a`` sets the number of hidden layers and their sizes before the embedding layer for user inputs;
```
Comment (Ghostvv) [inline docs/components.rst]: let's rephrase it to smith like: `sets a list of hidden layers sizes before embedding layer for user inputs, the number of hidden layers is equal to the length of the list`

File: docs/components.rst [hunk 2]
```diff
@@ -327,8 +327,8 @@ intent_classifier_tensorflow_embedding
...
+            - ``hidden_layers_sizes_b`` sets the number of hidden layers and their sizes before the embedding layer for intent labels;
```
Comment (Ghostvv) [inline docs/components.rst]: same as above

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 3]
```diff
@@ -396,6 +390,7 @@ def _create_batch_b(self, batch_pos_b, intent_ids):
         return np.concatenate([batch_pos_b, batch_neg_b], 1)
 
     def _linearly_increasing_batch_size(self, ep):
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: let's use the function from embedding policy, that checks whether `self.batch_size` is a `list`

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 4]
```diff
@@ -316,38 +306,41 @@ def _create_tf_embed_nn(self, x_in, is_training,
...
+            sim: between embedded words and embedded intent labels
+            sim_emb: between individual embedded intent labels only"""
 
         if self.similarity_type == 'cosine':
+            # normalize embedding vectors for cosine similarity
             a = tf.nn.l2_normalize(a, -1)
             b = tf.nn.l2_normalize(b, -1)
 
-        if self.similarity_type == 'cosine' or self.similarity_type == 'inner':
+        if self.similarity_type in ['cosine', 'inner']:
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: let's use set here `{...}`

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 5]
```diff
@@ -316,38 +306,41 @@ def _create_tf_embed_nn(self, x_in, is_training,
...
         else:
             raise ValueError("Wrong similarity type {}, "
                              "should be 'cosine' or 'inner'"
                              "".format(self.similarity_type))
 
     def _tf_loss(self, sim, sim_emb):
+        # type: (tf.Tensor, tf.Tensor) -> tf.Tensor
         """Define loss"""
 
         if self.use_max_sim_neg:
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: let's also refactor loss as in embedding policy, there is no need for `factors` tensor

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 6]
```diff
@@ -232,9 +151,78 @@ def __init__(self,
...
         self.intent_embed = intent_embed
 
+    # init helpers
+    def _load_nn_architecture_params(self, config):
+        # type: (Dict[Text, Any]) -> None
+        self.hidden_layer_sizes = {'a': config['hidden_layers_sizes_a'],
+                                   'b': config['hidden_layers_sizes_b']}
+
+        self.batch_size = config['batch_size']
+        if not isinstance(self.batch_size, list):
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: let's check for it in `_linearly_increasing_batch_size`

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 7]
```diff
@@ -405,6 +400,7 @@ def _linearly_increasing_batch_size(self, ep):
 
     def _train_tf(self, X, Y, intents_for_X,
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: parameters should have types

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 8]
```diff
@@ -296,14 +285,15 @@ def _prepare_data_for_training(self, training_data, intent_dict):
 
     # tf helpers:
     def _create_tf_embed_nn(self, x_in, is_training,
-                            num_layers, layer_size, name):
-        """Create embed nn for layer with name"""
+                            layer_sizes, name):
+        # type: (tf.Tensor, Bool, List, Text) -> tf.Tensor
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: I think `is_training` is a Tensor here, `List[int]`

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 9]
```diff
@@ -458,11 +454,12 @@ def _train_tf(self, X, Y, intents_for_X,
                 })
 
         if self.evaluate_on_num_examples:
-            logger.info("Finished training embedding policy, "
+            logger.info("Finished training embedding classifier, "
                         "loss={:.3f}, train accuracy={:.3f}"
                         "".format(last_loss, train_acc))
 
     def _output_training_stat(self, X, intents_for_X, is_training):
+        # type: (np.ndarray, np.ndarray, Bool) -> Float
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: `is_training` is a Tensor

File: rasa_nlu/classifiers/embedding_intent_classifier.py (hunk 3 above)
Comment (MetcalfeTom) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: Good catch, didn't realize this function was making unnecessary computations

File: rasa_nlu/classifiers/embedding_intent_classifier.py (hunk 7 above)
Comment (MetcalfeTom) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: I missed this one on purpose because it goes over two lines.  I'll update it, but the sight of it makes me feel ill

Comment (Ghostvv) [reply rasa_nlu/classifiers/embedding_intent_classifier.py]: use style of types as in `__init__`

**Comment by Ghostvv** [comment]
Let me know, when it is ready for a second review

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 10]
```diff
@@ -395,16 +384,29 @@ def _create_batch_b(self, batch_pos_b, intent_ids):
...
+                       epoch * (self.batch_size[1] -
+                                self.batch_size[0]) / (self.epochs - 1))
         else:
             return int(self.batch_size[0])
 
-    def _train_tf(self, X, Y, intents_for_X,
-                  loss, is_training, train_op):
+    def _train_tf(self,
+                  X,  # type: tf.Tensor
+                  Y,  # type: tf.Tensor
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: I think `X` and `Y` are `ndarray`
